File: textbook/python/01-dataflow/07_transactions.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — DataFlow / Transaction Management
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Use transaction management for atomic operations with
#            commit, rollback, and isolation level control
# LEVEL: Advanced
# PARITY: Full — Rust has DataFlowTransaction with RAII rollback
# VALIDATES: TransactionManager, transaction() context manager,
#            commit/rollback semantics, rollback_all(), get_active_transactions()
#
# Run: uv run python textbook/python/01-dataflow/07_transactions.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import os
import tempfile

from dataflow import DataFlow
from dataflow.features.transactions import TransactionManager

# ── Setup ───────────────────────────────────────────────────────────

db_path = os.path.join(tempfile.gettempdir(), "textbook_dataflow_07.db")
db = DataFlow(database_url=f"sqlite:///{db_path}")


@db.model
class Account:
    owner: str
    balance: float
    currency: str = "SGD"


async def main():
    await db.initialize()

    # Seed data
    await db.express.create(
        "Account", {"owner": "Alice", "balance": 1000.0, "currency": "SGD"}
    )
    await db.express.create(
        "Account", {"owner": "Bob", "balance": 500.0, "currency": "SGD"}
    )

    # ── 1. ACCESS — db.transactions property ────────────────────────
    # TransactionManager is accessed via db.transactions.

    txn_mgr = db.transactions
    assert isinstance(txn_mgr, TransactionManager)
    print("TransactionManager accessible via db.transactions")

    # ── 2. TRANSACTION — context manager for atomic operations ──────
    # transaction() yields a transaction context dict with:
    #   id, isolation_level, status, operations
    #
    # On normal exit: status becomes "committed"
    # On exception:   status becomes "rolled_back"

    with txn_mgr.transaction() as txn:
        assert txn["status"] == "active"
        assert txn["isolation_level"] == "READ_COMMITTED", "Default isolation"
        assert "id" in txn

        # Track operations inside the transaction
        txn["operations"].append({"type": "debit", "account": "Alice", "amount": 200})
        txn["operations"].append({"type": "credit", "account": "Bob", "amount": 200})

        print(f"Transaction {txn['id']}: {len(txn['operations'])} operations")

    # After the context manager exits normally, status is "committed"
    assert txn["status"] == "committed"
    print("Transaction committed successfully")

    # ── 3. ISOLATION LEVELS ─────────────────────────────────────────
    # Specify isolation level when creating a transaction.

    with txn_mgr.transaction(isolation_level="SERIALIZABLE") as txn:
        assert txn["isolation_level"] == "SERIALIZABLE"
        txn["operations"].append({"type": "audit", "details": "balance check"})

    assert txn["status"] == "committed"
    print("SERIALIZABLE transaction committed")

    # ── 4. ROLLBACK on exception ────────────────────────────────────
    # If an exception occurs inside the context manager, the transaction
    # is automatically rolled back.

    try:
        with txn_mgr.transaction() as txn:
            txn["operations"].append(
                {"type": "debit", "account": "Alice", "amount": 9999}
            )
            # Simulate a business rule violation
            raise ValueError("Insufficient funds")
    except ValueError:
        pass  # Expected

    assert txn["status"] == "rolled_back"
    assert txn.get("error") == "Insufficient funds"
    print("Transaction rolled back on exception")

    # ── 5. ACTIVE TRANSACTIONS — monitoring ─────────────────────────
    # get_active_transactions() returns currently active transactions.
    # After a transaction completes (commit or rollback), it is removed.

    active = txn_mgr.get_active_transactions()
    assert len(active) == 0, "No active transactions after all completed"

    # ── 6. ROLLBACK ALL — emergency operation ───────────────────────
    # rollback_all() rolls back every active transaction at once.
    # Useful for emergency shutdown or cleanup.

    result = txn_mgr.rollback_all()
    assert result["success"] is True
    assert result["count"] == 0, "No active transactions to roll back"
    assert isinstance(result["rolled_back_transactions"], list)
    print(f"Emergency rollback: {result['count']} transactions affected")

    # ── 7. Nested transaction pattern ───────────────────────────────
    # Transactions are independent — nesting creates separate contexts.

    with txn_mgr.transaction() as outer:
        outer["operations"].append({"type": "outer_op"})

        with txn_mgr.transaction() as inner:
            inner["operations"].append({"type": "inner_op"})

        # Inner committed, outer still active
        assert inner["status"] == "committed"

    assert outer["status"] == "committed"
    print("Nested transactions: both committed independently")

    # ── 8. Edge case: exception in nested transaction ───────────────
    # Inner rollback does not affect the outer transaction.

    with txn_mgr.transaction() as outer:
        outer["operations"].append({"type": "outer_op"})

        try:
            with txn_mgr.transaction() as inner:
                inner["operations"].append({"type": "risky_op"})
                raise RuntimeError("Inner failure")
        except RuntimeError:
            pass  # Inner rolled back

        assert inner["status"] == "rolled_back"
        # Outer continues
        outer["operations"].append({"type": "recovery_op"})

    assert outer["status"] == "committed"
    print("Inner rollback did not affect outer transaction")

    # ── 9. Transaction context data ─────────────────────────────────
    # The transaction context dict can carry arbitrary operation logs.

    with txn_mgr.transaction(isolation_level="READ_COMMITTED") as txn:
        txn["operations"].append(
            {
                "type": "transfer",
                "from": "Alice",
                "to": "Bob",
                "amount": 100.0,
                "currency": "SGD",
            }
        )

    assert len(txn["operations"]) == 1
    assert txn["operations"][0]["type"] == "transfer"
    assert txn["operations"][0]["amount"] == 100.0
    print("Transaction carries structured operation metadata")

    # ── Cleanup ─────────────────────────────────────────────────────
    await db.stop()


asyncio.run(main())

try:
    os.unlink(db_path)
except OSError:
    pass

print("PASS: 01-dataflow/07_transactions")

File: textbook/rust/01-dataflow/src/bin/07_transactions.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- DataFlow / Transactions
//!
//! OBJECTIVE: Use DataFlowTransaction for atomic multi-operation database changes.
//! LEVEL: Intermediate
//! PARITY: Full -- Python uses `async with db.transaction() as tx:`;
//!         Rust uses `df.begin_transaction().await` with explicit commit/rollback.
//!         RAII rollback ensures no partial writes on drop.
//! VALIDATES: DataFlowTransaction, begin_transaction, commit, rollback, RAII safety
//!
//! Run: cargo run -p tutorial-dataflow --bin 07_transactions

use kailash_dataflow::prelude::*;
use kailash_value::{Value, ValueMap};
use std::sync::Arc;

#[tokio::main]
async fn main() {
    // ── Setup ──
    let mut df = DataFlow::new("sqlite::memory:")
        .await
        .expect("connect");

    let model = ModelDefinition::new("Account", "accounts")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("name", FieldType::Text, |f| f.required())
        .field("balance", FieldType::Integer, |f| f.required())
        .auto_timestamps();

    df.register_model(model).expect("register");

    let express = df.express();

    // Create two accounts.
    let mut alice_data = ValueMap::new();
    alice_data.insert(Arc::from("name"), Value::from("Alice"));
    alice_data.insert(Arc::from("balance"), Value::Integer(1000));
    let alice = express
        .create("Account", alice_data)
        .await
        .expect("create alice");
    let alice_id = alice.get("id" as &str).unwrap().as_i64().unwrap();

    let mut bob_data = ValueMap::new();
    bob_data.insert(Arc::from("name"), Value::from("Bob"));
    bob_data.insert(Arc::from("balance"), Value::Integer(500));
    let bob = express
        .create("Account", bob_data)
        .await
        .expect("create bob");
    let bob_id = bob.get("id" as &str).unwrap().as_i64().unwrap();

    // ── 1. Successful Transaction: Transfer ──
    // Transfer 200 from Alice to Bob atomically.

    {
        let mut tx = df.begin_transaction().await.expect("begin tx");

        // Debit Alice.
        let mut debit = ValueMap::new();
        debit.insert(Arc::from("balance"), Value::Integer(800)); // 1000 - 200
        tx.update("Account", &alice_id.to_string(), debit)
            .await
            .expect("debit alice");

        // Credit Bob.
        let mut credit = ValueMap::new();
        credit.insert(Arc::from("balance"), Value::Integer(700)); // 500 + 200
        tx.update("Account", &bob_id.to_string(), credit)
            .await
            .expect("credit bob");

        // Commit makes both changes permanent.
        tx.commit().await.expect("commit transfer");
    }

    // Verify balances after successful transfer.
    let alice_after = express
        .read("Account", &alice_id.to_string())
        .await
        .expect("read alice");
    assert_eq!(
        alice_after.get("balance" as &str),
        Some(&Value::Integer(800))
    );

    let bob_after = express
        .read("Account", &bob_id.to_string())
        .await
        .expect("read bob");
    assert_eq!(
        bob_after.get("balance" as &str),
        Some(&Value::Integer(700))
    );

    // ── 2. RAII Rollback: Transaction Dropped Without Commit ──
    // If a transaction is dropped without calling commit(),
    // all changes are automatically rolled back.

    {
        let mut tx = df.begin_transaction().await.expect("begin tx");

        // Attempt to zero out Alice's balance.
        let mut zero = ValueMap::new();
        zero.insert(Arc::from("balance"), Value::Integer(0));
        tx.update("Account", &alice_id.to_string(), zero)
            .await
            .expect("zero alice");

        // tx is dropped here without commit() -- automatic rollback.
    }

    // Alice's balance is unchanged (still 800).
    let alice_check = express
        .read("Account", &alice_id.to_string())
        .await
        .expect("read alice after rollback");
    assert_eq!(
        alice_check.get("balance" as &str),
        Some(&Value::Integer(800))
    );

    // ── 3. Transaction with Create and Delete ──
    // Transactions can mix create, update, and delete operations.

    {
        let mut tx = df.begin_transaction().await.expect("begin tx");

        let mut new_account = ValueMap::new();
        new_account.insert(Arc::from("name"), Value::from("Charlie"));
        new_account.insert(Arc::from("balance"), Value::Integer(300));
        tx.create("Account", new_account)
            .await
            .expect("create charlie");

        tx.commit().await.expect("commit create");
    }

    let count = express.count("Account").await.expect("count");
    assert_eq!(count, 3); // Alice, Bob, Charlie

    println!("PASS: 01-dataflow/07_transactions");
}
